homeassistant/components/intent/timers.py

```diff
@@ -0,0 +1,772 @@
+        start_hours = int(slots["start_hours"]["value"])
+
+    start_minutes: int | None = None
+    if "start_minutes" in slots:
+        start_minutes = int(slots["start_minutes"]["value"])
+
+    start_seconds: int | None = None
+    if "start_seconds" in slots:
+        start_seconds = int(slots["start_seconds"]["value"])
+
+    if (
+        (start_hours is not None)
+        or (start_minutes is not None)
+        or (start_seconds is not None)
+    ):
+        has_filter = True
+        matching_timers = [
+            t
+            for t in matching_timers
+            if (t.start_hours == start_hours)
+            and (t.start_minutes == start_minutes)
+            and (t.start_seconds == start_seconds)
+        ]
+
+        if len(matching_timers) == 1:
+            # Only 1 match remaining
+            return matching_timers[0]
+
+    if (not has_filter) and (len(matching_timers) == 1):
+        # Only 1 match remaining with no filter
+        return matching_timers[0]
+
+    # Use device id
+    device_id: str | None = None
+    if matching_timers and ("device_id" in slots):
+        device_id = slots["device_id"]["value"]
+        assert device_id is not None
+        matching_device_timers = [
+            t for t in matching_timers if (t.device_id == device_id)
+        ]
+        if len(matching_device_timers) == 1:
+            # Only 1 match remaining
+            return matching_device_timers[0]
+
+        # Try area/floor
+        if (
+            (device := timer_manager.device_registry.async_get(device_id))
+            and device.area_id
+            and (area := timer_manager.area_registry.async_get_area(device.area_id))
+        ):
+            # Try area
+            matching_area_timers = [
+                t for t in matching_timers if (t.area_id == area.id)
+            ]
+            if len(matching_area_timers) == 1:
+                # Only 1 match remaining
+                return matching_area_timers[0]
+
+            # Try floor
+            matching_floor_timers = [
+                t for t in matching_timers if (t.floor_id == area.floor_id)
+            ]
+            if len(matching_floor_timers) == 1:
+                # Only 1 match remaining
+                return matching_floor_timers[0]
+
+    if has_filter and ("ordinal" in slots):
+        ordinal = int(slots["ordinal"]["value"])
+        if 0 <= ordinal < len(matching_timers):
+            # Sort by creation time
+            sorted_timers = sorted(matching_timers, key=lambda t: t.created_at)
+            return sorted_timers[ordinal]
+
+    if matching_timers:
+        raise MultipleTimersMatchedError
+
+    _LOGGER.warning(
+        "Timer not found: name=%s, hours=%s, minutes=%s, seconds=%s, device_id=%s",
+        name,
+        start_hours,
+        start_minutes,
+        start_seconds,
+        device_id,
+    )
+
+    raise TimerNotFoundError
+
+
+def _find_timers(hass: HomeAssistant, slots: dict[str, Any]) -> list[TimerInfo]:
```

Comment:
this logic is duplicate? Why not have `find_timer` wrap `find_timers` and just raise if it returns 0 or 2+ timers ?

Reply:
The early exit condition is different: `find_timer` exits early if only a single timer remains whereas `find_timer` exits when no timers remain. I can still combine them with an extra check, though.